### DISCUSSION TITLE
Read/write separation

### DISCUSSION THREAD

**Discussion by dtdionne** [solidusio/solidus#5732] [Troubleshooting] [answered]
I’m still researching this but I figured I’d ask.  Does or can solidus easily support a high availability Postgres patroni cluster?  I’ve got a little cluster that can read from replicas on port 12000 and read/write to the master on port 12001.

I’m not a developer so please forgive me if this is a silly question.

**Comment by kennyadsl** [comment]
@dtdionne Solidus is basically a Rails application, so you can configure multiple database within a Solidus application. I suggest to study this page of the Rails documentation, which also apply to the Active Record classes (all the models) we have in Solidus: https://guides.rubyonrails.org/active_record_multiple_databases.html.

**Reply by dtdionne** [reply]
Yeah I skimmed that one yesterday and I wasn’t sure it was applicable because it appeared that an intimate developer level knowledge of the application was required to implement.  I don’t know if that’s true.  But if so, I don’t have this knowledge.  So that’s kinda what prompted me to broach this here.

**Reply by dtdionne** [reply]
Unless rails can obfuscate that, which would be great.  I can test that pretty easily.

**Reply by jarednorman** [reply]
Given that Solidus long predates support for multiple databases in Rails, it doesn't make some guarantees that can make doing this easier. For example, you can't determine the whether a requests queries should go to the read or write instance based on the HTTP method. That said, with a little work you can absolutely leverage multiple databases and I've done it on a number of stores.

**Reply by dtdionne** [reply]
Gadzooks!!!  I think it just works!?!?  Can that be so?

I added a primary and replica to my database.yml per that doc.  Primary pointing to the read/write port and replica pointing to the read only port.  The app started and works.  I guess I’ll have to try and verify that reads are going to the read only port somehow.

**Reply by dtdionne** [reply]
Ohhh, so the application must be massaged to know when to use the primary or replica?  I was hoping that rails could abstract that away.

**Reply by jarednorman** [reply]
Rails gives you the ability to control when to use each and provides a basic implementation that will route the requests based on HTTP method, but that only works if your site never writes on GET/HEAD requests, which I suspect isn't the case for a few routes on a Solidus store. Certainly worth a try though.

**Reply by dtdionne** [reply]
But I can still do a quick and dirty test with tcpdump to see if any traffics happening on the read only port.  I’ll first try without the app running to get a baseline then crank up the app and browse around and see if there’s traffic on the read port.

**Reply by dtdionne** [reply]
Hey Jared, thanks for discussing this with me, I really appreciate it :)

**Reply by dtdionne** [reply]
I’m lost in a world I don’t understand and that kinda scares me…so thank you.

**Reply by dtdionne** [reply]
Yeah nothing went over the read port.  But it did the read/write port.  Double dukes.

**Reply by dtdionne** [reply]
We tried adding to application_record.rb as that doc explains but it doesn’t appear the app uses that file.  I deleted it as a test and the app still appears to run fine.

**Reply by dtdionne** [reply]
I'm sorry, i don't know what GET/HEAD means nor exactly how to test for it.  A rails routes -E | grep -i head doesnt return any hits...but idk what that means.

**Reply by jarednorman** [reply, marked as answer]
HTTP requests have a [method](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods). Don't worry to much that there are specific HEAD routes in the app. That's normal and not worth spending time worrying about.

I will offer the friendly advice that if you are not pretty familiar with Rails development specifically and web development in general, operating a Solidus store will likely be very difficult for you. Getting multi-DB support working on a complex app like a Solidus store is not for the faint of heart. It would be cool if it was something we officially supported, but that's not currently the case. All the stores I've worked that have support rolled their own and it's optimized for their specific use cases (like offloading certain very expensive queries to the read replicas.)

**Reply by dtdionne** [reply]
Thanks Jared
